app/src/main/java/com/nextcloud/talk/remotefilebrowser/activities/RemoteFileBrowserActivity.kt

```diff
@@ -0,0 +1,343 @@
+/*
+ * Nextcloud Talk application
+ *
+ * @author Andy Scherzinger
+ * Copyright (C) 2022 Andy Scherzinger <info@andy-scherzinger.de>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.nextcloud.talk.remotefilebrowser.activities
+
+import android.app.Activity
+import android.content.Intent
+import android.os.Bundle
+import android.util.Log
+import android.view.Menu
+import android.view.MenuItem
+import android.view.View
+import androidx.appcompat.app.AppCompatActivity
+import androidx.core.content.res.ResourcesCompat
+import androidx.fragment.app.DialogFragment
+import androidx.lifecycle.ViewModelProvider
+import androidx.recyclerview.widget.GridLayoutManager
+import androidx.recyclerview.widget.LinearLayoutManager
+import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
+import autodagger.AutoInjector
+import com.nextcloud.talk.R
+import com.nextcloud.talk.application.NextcloudTalkApplication
+import com.nextcloud.talk.databinding.ActivityRemoteFileBrowserBinding
+import com.nextcloud.talk.interfaces.SelectionInterface
+import com.nextcloud.talk.remotefilebrowser.adapters.RemoteFileBrowserItemsAdapter
+import com.nextcloud.talk.remotefilebrowser.model.RemoteFileBrowserItem
+import com.nextcloud.talk.remotefilebrowser.viewmodels.RemoteFileBrowserItemsViewModel
+import com.nextcloud.talk.ui.dialog.SortingOrderDialogFragment
+import com.nextcloud.talk.utils.DisplayUtils
+import com.nextcloud.talk.utils.FileSortOrder
+import com.nextcloud.talk.utils.FileSortOrderNew
+import com.nextcloud.talk.utils.database.user.UserUtils
+import com.nextcloud.talk.utils.preferences.AppPreferences
+import net.orange_box.storebox.listeners.OnPreferenceValueChangedListener
+import java.io.File
+import java.util.Collections
+import java.util.TreeSet
+import javax.inject.Inject
+
+@AutoInjector(NextcloudTalkApplication::class)
+class RemoteFileBrowserActivity : AppCompatActivity(), SelectionInterface, SwipeRefreshLayout.OnRefreshListener {
```

Comment:
This activity contains some data and logic that should be in the viewmodel (and some of it exposed through the state livedata):

- `browserItems` and sorting logic
- `selectedPaths`, `currentPath` and selection logic

In general, if the activity/fragment is holding model data, then something's wrong. The activity should strictly only transform data into views and pass events back to the viewmodel.
If you feel that this is too much data for a single State, feel free to split it (one for items/load and another for selection, for example)

Reply:
I lack the knowledge or idea(s) how to move this to the viewmodel, especially the part about selected and current path due to the fact that:
* selected path is a collection holding various, different path sections from the whole file tree while the `browserItems` collection is always just reflecting the current path, hence my struggle

as for the state

* I just adapted the state logic from the initial example from the shared items implementation not truly understanding the state concept :/

So happy for any hints and pointers 🙏 or help with the changes

Reply:
@AlvaroBrey Any idea how to do this? I simply lack the knowledge. Most examples on the internet do the sorting via the repositoy which seems strange since that mean I need to do a server round trip just to re-sort an existing result which might be okay to do it is is a database but feel very wrong for remote calls.
Also given the current implementation I have no access to the result (file list) since it is directly wrapped into the view state adding another issue: being able to change the sorting while the file listing might be empty or loading. Again any help and pointers are appreciated ❤️

Reply:
> Most examples on the internet do the sorting via the repositoy which seems strange since that mean I need to do a server round trip just to re-sort an existing result which might be okay to do it is is a database but feel very wrong for remote calls. 

Most examples use server-side sorting because they also use server-side pagination, so if sort is changed the already retrieved pages are invalid anyway. I'll take a look and see how it can be improved here.

Reply:
also done for current path, selected paths

Reply:
❤️ Thanks a lot @AlvaroBrey - will have a look to learn how to do this myself in the future 🔮